node/src/components/block_proposer.rs

```diff
@@ -3,189 +3,273 @@
+    sets: BlockProposerDeploySets,
+    // We don't need the whole Chainspec here, just the deploy config.
+    deploy_config: DeployConfig,
+    /// Key for storing the block proposer state.
+    state_key: Vec<u8>,
+    /// The queue of requests awaiting being handled.
+    request_queue: RequestQueue,
 }
 
-impl BlockProposer {
-    /// Creates a new, block proposer instance with the provided internal state.
-    pub(crate) fn new<REv>(
-        registry: Registry,
+impl BlockProposerReady {
+    fn handle_event<REv>(
+        &mut self,
         effect_builder: EffectBuilder<REv>,
-        state: BlockProposerState,
-    ) -> Result<(Self, Effects<Event>), prometheus::Error>
+        event: Event,
+    ) -> Effects<Event>
     where
-        REv: ReactorEventT,
+        REv: Send + From<StateStoreRequest>,
     {
-        let effects = effect_builder
-            .set_timeout(PRUNE_INTERVAL)
-            .event(|_| Event::BufferPrune);
-
-        let metrics = BlockProposerMetrics::new(registry)?;
-        let this = BlockProposer {
-            metrics,
-            state,
-            chainspecs: HashMap::new(),
-        };
-        Ok((this, effects))
+        match event {
+            Event::Request(BlockProposerRequest::ListForInclusion(request)) => {
+                if request.next_finalized > self.sets.next_finalized {
+                    self.request_queue
+                        .entry(request.next_finalized)
+                        .or_default()
+                        .push(request);
+                    Effects::new()
+                } else {
+                    request
+                        .responder
+                        .respond(self.propose_deploys(
+                            self.deploy_config,
+                            request.current_instant,
+                            request.past_deploys,
+                        ))
+                        .ignore()
+                }
+            }
+            Event::Buffer { hash, header } => {
+                self.add_deploy(Timestamp::now(), hash, *header);
+                Effects::new()
+            }
+            Event::BufferPrune => {
+                let pruned = self.prune(Timestamp::now());
+                log::debug!("Pruned {} deploys from buffer", pruned);
+
+                // After pruning, we store a state snapshot.
+                let mut effects = effect_builder
+                    .save_state(self.state_key.clone().into(), self.sets.clone())
+                    .ignore();
+
+                // Re-trigger timer after `PRUNE_INTERVAL`.
+                effects.extend(
+                    effect_builder
+                        .set_timeout(PRUNE_INTERVAL)
+                        .event(|_| Event::BufferPrune),
+                );
+
+                effects
+            }
+            Event::Loaded { sets, .. } => {
+                // This should never happen, but we can just ignore the event and carry on.
+                warn!(
```

Comment:
I'd be inclined to make this an `error!` since it's an indicator that something has gone very wrong.

Reply:
I'd prefer to keep this a `warn!` - if it occurs, it is strange and a bug, but we likely can carry on going with little harm.

`error!` is reserved for things where someone would get notified to work on fixing the bug or the node deployment right away.